src/menelaus/ensemble.py

```diff
@@ -77,64 +87,113 @@ def eval_confirmed_approval(approvals_needed=1, confirmations_needed=1):
-            if det.drift_state == "drift":
-
-                if num_approvals < approvals_needed:
+        for d in detectors:
+            if d.drift_state == "drift":
+                if num_approvals < self.approvals_needed:
                     num_approvals += 1
                 else:
                     num_confirmations += 1
 
                 if (
-                    num_approvals >= approvals_needed
-                    and num_confirmations >= confirmations_needed
+                    num_approvals >= self.approvals_needed
+                    and num_confirmations >= self.confirmations_needed
                 ):
                     return "drift"
 
         return None
 
-    return f
 
+class MacielElection(Election):
+    """
+    Maciel election for handling detectors (typically in
+    streaming setting) with waiting logic.
+    """
+
+    def __init__(self, sensitivity: int, wait_time: int):
+        """
+        Args:
+            sensitivity (int): how many combined waiting/new drift alarms
+                should result in ensemble alarm
+            wait_time (int): after how many steps of waiting, should each
+                detector reset its time spent waiting post-drift-alarm
+        """
+        self.sensitivity = sensitivity
+        self.wait_time = wait_time
+        self.wait_period_counters = None
+
+    def __call__(self, detectors: list):
+        """
+        Args:
+            detectors (list): detector objects to examine
 
-# TODO - How to put n-min-approvals version of evaluator
-#        in table, since the user may change n? Same with others.
-# TODO - Isn't confirmed approval just a version of minimum
-#        approval, with a non-guaranteed order characteristic?
-EVALUATORS = {
-    "simple-majority": eval_simple_majority,
-    "single-approval": eval_minimum_approval(1),
-    "single-confirmed-approval": eval_confirmed_approval(1, 1),
-}
+        Returns:
+            str: ``"drift_state"`` if drift is determined, or ``None``
+        """
+        # XXX - AS thinks it is ok for this to be defined here rather
+        #       than init. At first evaluation (i.e. first update),
+        #       the list of 0s is made. Afterwards, it can be updated.
+        #       The detectors are passed-by-ref, so they should be
+        #       up-to-date upon each call.
+        if self.wait_period_counters is None:
+            self.wait_period_counters = [0 for _ in detectors]
+
+        num_drift = 0
+        num_warning = 0
+
+        states = [d.drift_state for d in detectors]
+        for i, state in enumerate(states):
+            if state == "drift" and self.wait_period_counters[i] == 0:
+                num_drift += 1
+                self.wait_period_counters[i] += 1
+            elif state == "warning":
+                num_warning += 1
+            elif self.wait_period_counters[i] != 0:
+                # XXX - So the same detector waiting will increment num_drift? - AS
```

Comment:
Since this is an `elif`, this should only execute after the first `update` where `drift_state == 'drift'`.